refactor(parser): use checkpoints instead of `peek_at` in `is_un_parenthesized_async_arrow_function_worker`

camchenry opened the pull request:
- towards https://github.com/oxc-project/oxc/issues/11194

Replaces `peek_at` etc. with checkpoints and rewind, so that we don't need token lookaheads in the future. This is also more similar to how the TypeScript compiler parses these expressions.

camchenry commented:
* **#11218** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11218?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11218?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11218).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
Really interesting this doesn't regress performance at all ... probably due to no usage in the benchmarks.

camchenry commented:
> Really interesting this doesn't regress performance at all ... probably due to no usage in the benchmarks.

Yeah, I'm guessing this is exercised so rarely, I'd expect the following PR to have a larger regression possibly.

camchenry commented on crates/oxc_parser/src/js/arrow.rs:
```diff
@@ -190,20 +190,24 @@ impl<'a> ParserImpl<'a> {
-            let first = first_token.kind();
+            let checkpoint = self.checkpoint();
+            self.bump(Kind::Async);
             // If the "async" is followed by "=>" token then it is not a beginning of an async arrow-function
             // but instead a simple arrow-function which will be parsed inside "parseAssignmentExpressionOrHigher"
-            if first_token.is_on_new_line() || first == Kind::Arrow {
+            if self.cur_token().is_on_new_line() || self.at(Kind::Arrow) {
+                self.rewind(checkpoint);
                 return Tristate::False;
             }
             // Check for un-parenthesized AsyncArrowFunction
-            if first.is_binding_identifier() {
+            if self.cur_kind().is_binding_identifier() {
                 // Arrow before newline is checked in `parse_simple_arrow_function_expression`
-                if self.nth_at(2, Kind::Arrow) {
+                self.bump_any();
+                if self.at(Kind::Arrow) {
+                    self.rewind(checkpoint);
                     return Tristate::True;
                 }
             }
+            self.rewind(checkpoint);
         }
         Tristate::False
```
going to leave this as-is, for easier comparison with the TS compiler

camchenry committed:
refactor(parser): use checkpoints instead of `peek_at` in `is_un_parenthesized_async_arrow_function_worker` (#11218)

- towards https://github.com/oxc-project/oxc/issues/11194

Replaces `peek_at` etc. with checkpoints and rewind, so that we don't need token lookaheads in the future. This is also more similar to how the TypeScript compiler parses these expressions.
